internal/pkg/deploy/cloudformation/stack/env.go

```diff
@@ -243,31 +244,57 @@ func (e *EnvStackConfig) transformParameters(
 	// Remove or transform each of the current parameters.
 	var params []*cloudformation.Parameter
 	for k, p := range curr {
-		if transformed := transform(p, old[k]); transformed != nil {
-			params = append(params, transformed)
+		currP := p
+		for _, transform := range transformFunc {
+			currP = transform(currP, old[k])
+		}
+		if currP != nil {
+			params = append(params, currP)
 		}
 	}
 	return params, nil
 }
 
-// transformEnvControllerParameters transforms a parameter such that it uses its previous value if:
-// 1. The parameter exists in the old template.
-// 2. The parameter is env-controller managed.
-// Otherwise, it returns the parameter untouched.
-func transformEnvControllerParameters(new cloudformation.Parameter, old *cloudformation.Parameter) *cloudformation.Parameter {
-	if old == nil { // The ParamKey doesn't exist in the old stack, use the new value.
-		return &new
+// transformEnvControllerParameters transforms an env-controller managed parameter.
+// If the parameter exists in the old template, it uses its previous value. Otherwise, it returns its new default value.
+func transformEnvControllerParameters(new, old *cloudformation.Parameter) *cloudformation.Parameter {
+	if new == nil {
+		return nil
 	}
-
-	var (
-		isEnvControllerManaged = make(map[string]struct{})
-		exists                 = struct{}{}
-	)
+	isEnvControllerManaged := make(map[string]struct{})
 	for _, f := range template.AvailableEnvFeatures() {
-		isEnvControllerManaged[f] = exists
+		isEnvControllerManaged[f] = struct{}{}
 	}
 	if _, ok := isEnvControllerManaged[aws.StringValue(new.ParameterKey)]; !ok {
-		return &new
+		return new
+	}
+	if old == nil { // The EnvController-managed parameter doesn't exist in the old stack. Use the new value.
+		return new
+	}
+	return &cloudformation.Parameter{
+		ParameterKey: new.ParameterKey,
+
+		// Ideally, we would set `UsePreviousValue: true` unfortunately CodePipeline template config does not support it.
+		// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/continuous-delivery-codepipeline-cfn-artifacts.html#w2ab1c21c15c15
+		ParameterValue: old.ParameterValue,
+	}
+}
+
+// transformServiceDiscoveryEndpoint transforms the service discovery endpoint parameter.
+// If the parameter exists in the old template, it uses its previous value.
+// Otherwise, it uses a default value of `<app>.local`.
+func (e *EnvStackConfig) transformServiceDiscoveryEndpoint(new, old *cloudformation.Parameter) *cloudformation.Parameter {
+	if new == nil {
+		return nil
+	}
+	if aws.StringValue(new.ParameterKey) != EnvParamServiceDiscoveryEndpoint {
+		return new
+	}
+	if old == nil {
+		return &cloudformation.Parameter{
+			ParameterKey:   new.ParameterKey,
+			ParameterValue: aws.String(fmt.Sprintf(`%s.local`, e.in.App.Name)),
```

Comment:
wait, isn't the new default `env.app.local`?

Reply:
Yes but we want to use the old default in this case. Otherwise we'd be trying to update `app.local` to `env.app.local` which is the bug in the first place

Reply:
Ah, gotcha gotcha-- I missed the second part of "the fix."